From cf6202e6e2a7200c4590cae1d2502bc8e1c2c95d Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Markus=20M=C3=BCtzel?= <markus.muetzel@gmx.de>
Date: Thu, 15 Oct 2026 11:38:55 +0200
Subject: [PATCH] [flang] Fix semantics check regarding OpenMP simd constructs.

There is an error in the wording of the OpenMP 5.2 specification.
LLVM Flang implemented a semantics check following that erroneous wording.

Fix that check to avoid regressions in projects that are using these simd
constructs.

See also: https://github.com/llvm/llvm-project/pull/177055
---
 lib/Semantics/check-omp-loop.cpp | 6 +++---
 1 file changed, 3 insertions(+), 3 deletions(-)

diff --git a/lib/Semantics/check-omp-loop.cpp b/lib/Semantics/check-omp-loop.cpp
index 5165ca5f70d0..36f28ea315e9 100644
--- a/lib/Semantics/check-omp-loop.cpp
+++ b/lib/Semantics/check-omp-loop.cpp
@@ -766,10 +766,10 @@
           if (linearMod->v != parser::OmpLinearModifier::Value::Ref) {
             CheckIntegerNoRef(symbol, source);
           } else {
-            if (!IsAllocatable(*symbol) && !IsAssumedShape(*symbol) &&
-                !IsPolymorphic(*symbol)) {
+            if (IsAllocatable(*symbol) || IsAssumedShape(*symbol) ||
+                IsPolymorphic(*symbol)) {
               context_.Say(source,
-                  "The list item `%s` specified with the REF '%s' must be polymorphic variable, assumed-shape array, or a variable with the `ALLOCATABLE` attribute"_err_en_US,
+                  "The list item `%s` specified with the REF '%s' must not be a polymorphic variable, an assumed-shape array, or a variable with the `ALLOCATABLE` attribute"_err_en_US,
                   symbol->name(), desc.name.str());
             }
           }
-- 
2.51.0.windows.2

